### ISSUE TITLE
The task$row_roles$use List Should Be Modifiable Or Replaceable

### ISSUE THREAD

**Issue by drag05** [mlr-org/mlr3#1302]
`mlr3` version 0.23.0
I am running a classification task involving stratified sampling on one stratum:
```
task =  TaskClassif$new('prep', target = 'type', extra_args = list(roles = 'allways_included')
                        , backend = DataBackendDataTable$new(dt, primary_key = 'id'))

> task$col_roles$stratum
[1] "charge"

```
Data columns 'type' and 'charge' are respectively factor and integer.

Partitioning of the task, results in

```
tsplit = partition(task, ratio = c(0.77, 0.13))

> names(tsplit)
[1] "train"      "test"       "validation"

> lengths(tsplit)
     train       test validation 
    807402     136315     104856 
```

 To avoid carrying the result of partition  (similar with `resampling`) downstream, I find more advantageous having these row id roles and indices contained by the task itself, either through the `set_row_roles` function or other method so that further training, testing and validation of pipeops, learners etc. be run on these subsets without overhead code (e.g. just by `learner$train(task)`). 

This would  require that the "use" list mutates easily into "train", "test" and "validation" lists of indices. 

Currently, using function `task$set_row_roles()` in two forms throws the same error:

```
lapply(
         seq.int(tsplit), \(i) task$set_row_roles(
                                                rows = tsplit[[i]]
                                              , roles = names(tsplit)[i]
                                              , remove_from = task$row_roles['use']
                    )
)

````
Or simpler, as suggested by the plural nouns used in arguments:

```
task$set_row_roles(rows = c(tsplit[[1]], tsplit[[2]], tsplit[[3]]), roles = names(tsplit), ...) 

Error in task_set_roles(private$.row_roles, rows, roles, add_to, remove_from,  : 
  Assertion on 'roles' failed: Must be a subset of {'use'}, but has additional elements {'train'}.
```
regardless of which of the `remove_from` or `add_to` arguments is used.

It seems that the list "use" is immutable although the presence of function `set_row_roles()` suggests otherwise.  

On a similar topic: setting the target variable as stratum, as suggested in Example at https://mlr3.mlr-org.com/reference/Resampling.html now throws an error:

```
> task =  TaskClassif$new('prep', target = 'type', ...)
 
> task$set_col_roles(cols = 'type', roles = 'stratum')

Error: Supervised tasks need at least one target column
```
which would suggest that the target cannot be used as stratum (contrary to the Example linked above) and that stratified split relies on the existence of - at least - another, lower cardinality (grouping) variable present in the data. 

Please advise, thank you!

**Comment by sebffischer** [comment]
Hi @drag05, thanks for opening this issue!

However, I am fairly certain that will *not* extend the `$row_roles`. In fact, other roles have existed in the past, but because the whole mechanism was somewhat confusing, we have removed them.

Maybe you can elaborate a bit more on where there is the "overhead code"?

Also, `resample()` already covers the use case of training the learner and making predictions on specific sets, so it's not clear to me why we also need `learner$train(task)` support this as well. 

You can specify a custom resampling and set a validation task, see the code further below.
I would agree that this is not as beautiful as we would like it to be (which his because the whole validation mechanism was added after the initial design of `mlr3`), but maybe it does what you want?
You can also read more about this topic in [chapter 15 of the mlr3 book](https://mlr3book.mlr-org.com/chapters/chapter15/predsets_valid_inttune.html). Let me know whether the chapter is missing some important information.

``` r
library(mlr3)

task = tsk("iris")

split = partition(task, c(0.6, 0.2))

task$internal_valid_task = split$validation
task$internal_valid_task
#> <TaskClassif:iris> (20 x 5): Iris Flowers
#> * Target: Species
#> * Properties: multiclass
#> * Features (4):
#>   - dbl (4): Petal.Length, Petal.Width, Sepal.Length, Sepal.Width

learner = lrn("classif.debug")

resampling = rsmp("custom")$instantiate(task, 
  train_sets = list(split$train),
  test_sets= list(split$test)
)

learner$predict_sets = c("train", "test", "internal_valid")

rr = resample(task, learner, resampling, store_models = TRUE)
#> INFO  [11:21:48.588] [mlr3] Applying learner 'classif.debug' on task 'iris' (iter 1/1)

rr$predictions("test")
#> [[1]]
#> <PredictionClassif> for 20 observations:
#>  row_ids     truth response
#>        2    setosa   setosa
#>        4    setosa   setosa
#>       15    setosa   setosa
#>      ---       ---      ---
#>      142 virginica   setosa
#>      145 virginica   setosa
#>      146 virginica   setosa
rr$predictions("train")
#> [[1]]
#> <PredictionClassif> for 60 observations:
#>  row_ids     truth response
#>        1    setosa   setosa
#>        5    setosa   setosa
#>        7    setosa   setosa
#>      ---       ---      ---
#>      139 virginica   setosa
#>      141 virginica   setosa
#>      148 virginica   setosa
rr$predictions("internal_valid")
#> [[1]]
#> <PredictionClassif> for 20 observations:
#>  row_ids     truth response
#>       21    setosa   setosa
#>       23    setosa   setosa
#>       31    setosa   setosa
#>      ---       ---      ---
#>      126 virginica   setosa
#>      130 virginica   setosa
#>      132 virginica   setosa
```

<sup>Created on 2025-05-06 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

One thing that we might want to think about is whether we want to support resamplings generating not only the "train" and "test" set, but also the "validation" data.

Then we code support something like the following, which seems more user-friendly to me.

```r
split = partition(task, c(0.6, 0.2))
resampling = rsmp("custom")$instantiate(task,
  train_sets = list(split$train),
  valid_sets = list(split$train),
  test_sets = list(split$train)
)
```

**Comment by sebffischer** [comment]
Note that to set the desired strata as follows, you can use the `add_to` argument of `$set_col_roles()`.

``` r
task = tsk("iris")
task$set_col_roles("Species", add_to = "stratum")
task
#> <TaskClassif:iris> (150 x 5): Iris Flowers
#> * Target: Species
#> * Properties: multiclass, strata
#> * Features (4):
#>   - dbl (4): Petal.Length, Petal.Width, Sepal.Length, Sepal.Width
#> * Strata: Species
```

Alternatively, you can specify all roles, i.e. `roles = c("target", "stratum")`:

``` r
task = tsk("iris")
task$set_col_roles("Species", roles = c("stratum", "target"))
task
#> <TaskClassif:iris> (150 x 5): Iris Flowers
#> * Target: Species
#> * Properties: multiclass, strata
#> * Features (4):
#>   - dbl (4): Petal.Length, Petal.Width, Sepal.Length, Sepal.Width
#> * Strata: Species
```

**Comment by drag05** [comment]
@sebffischer 
Thank you for taking the time! I hope you understand my confusion due to my recent exposure to mlr3 and the fact that I find parts of the documentation vague or more dev-oriented than user-oriented or contradictory, which is understandable to some extent as work is still in progress. 

Regarding the above example involving the "iris" task: after resampling, the task 

- contains redundant information regarding the "training" rows
```
identical(task$row_ids, task$row_roles$use)
[1] TRUE
```
- while it contains no "test" rows (they remain inside `rr`) 
- contains "validation" rows inserted as... internal task (?) while the presence of `set_row_roles()` suggests a more logical course
- unclear if, after resampling, the learner is trained (as in `learner$is_trained TRUE/FALSE`)  since the predictions are carried on `rr` and not on learner itself

From my experience, the purpose of `cv` etc. resampling  is minimization of overfit/bias the result being the learner trained. 
Here, the result is a new object (i.e. `rr`) containing the (trained?) learner. This is the overhead code (actually objects) that need to be carried downstream. 

The flow should output two objects only: the task containing split indices and the trained learner:

- resample for said learner(s)
- store train, test, validation indices in task
- obtain trained learner(s) through above resampling
- validate the learner(s) e.g. `learner$validate(task)` or `learner$predict(task, row_ids = 'validation')`
- use the `learner(s)` to predict on new data and not other objects such as `rr`
- eventually repeat the cycle on new data when prediction performance drops

I think `feature selection` does a better job as the selected features are easily inserted into the task. Row selection following a similar pattern would add consistency to the entire process. Thank you!

**Comment by sebffischer** [comment]
Are there other parts of the documentation (besides the row roles) that are hard to understand/inconsistent? If you could point to those places, we would love to improve them.

Yes, we should probably remove `$row_roles` and `$set_row_roles` as these are the remains of a previous mechanism that was not fully thought through. I can understand that this is confusing but it is extremely unlikely that we will add support for other row roles again. This is not because we don't like the idea of having a simple mechanism, but this was not working out in conjunction with preprocessing in `mlr3pipelines` (because there, different rows of "the same task" are preprocessed differently, depending on whether they are in the train, validation, or test set).

> From my experience, the purpose of cv etc. resampling is minimization of overfit/bias the result being the learner trained.

This is incorrect. Resampling can also simply be used for performance evaluation.

Note that you can extract the trained learner from a resample result:

``` r
library(mlr3)

task = tsk("iris")

rr = resample(task, lrn("classif.rpart"), rsmp("holdout"), store_models = TRUE)
#> INFO  [20:15:24.208] [mlr3] Applying learner 'classif.rpart' on task 'iris' (iter 1/1)

learner = rr$learners[[1]]
learner$model
#> n= 100 
#> 
#> node), split, n, loss, yval, (yprob)
#>       * denotes terminal node
#> 
#> 1) root 100 65 versicolor (0.33000000 0.35000000 0.32000000)  
#>   2) Petal.Length< 2.6 33  0 setosa (1.00000000 0.00000000 0.00000000) *
#>   3) Petal.Length>=2.6 67 32 versicolor (0.00000000 0.52238806 0.47761194)  
#>     6) Petal.Width< 1.55 37  3 versicolor (0.00000000 0.91891892 0.08108108) *
#>     7) Petal.Width>=1.55 30  1 virginica (0.00000000 0.03333333 0.96666667) *
```

I am also confused what you mean with

> eventually repeat the cycle on new data when prediction performance drops

do you want to perform hyperparameter optimization?

I can really recommend you reading our book, which hopefully removes some of your confusion: https://mlr3book.mlr-org.com/

**Comment by drag05** [comment]
> Are there other parts of the documentation (besides the row roles) that are hard to understand/inconsistent? If you could point to those places, we would love to improve them.

I will do my best.

> Yes, we should probably remove $row_roles and $set_row_roles as these are the remains of a previous mechanism that was not fully thought through.

If this was documented somewhere, I did not find it.

>This is incorrect. Resampling can also simply be used for performance evaluation.

There is a logical inconsistency in your statement: "is incorrect" along with "it can also". I agree with your end of statement as training performance is used as a measure for training evaluation.

>Note that you can extract the trained learner from a resample result

So, 
```
> learner = rr$learners[[1]]
```
is a trained learner

```
> learner$predict(task$internal_valid_task)
<PredictionClassif> for 30 observations:
 row_ids     truth response
       2    setosa   setosa
       7    setosa   setosa
       9    setosa   setosa
     ---       ---      ---
     144 virginica   setosa
     148 virginica   setosa
     149 virginica   setosa
```

>I am also confused what you mean with

>>eventually repeat the cycle on new data when prediction performance drops

I was referring to production environment. Never mind! 

>I can really recommend you reading our book

The version of the book I am reading is probably outdated. I'll check. Thank you!

**Comment by drag05** [comment]
@sebffischer Thank you for your time!
